plugin-server/src/worker/ingestion/event-property-counter.ts

```diff
@@ -0,0 +1,178 @@
+                for (const [key, propertyBuffer] of teamBuffer.entries()) {
+                    const [event, property] = JSON.parse(key)
+                    const { propertyType, propertyTypeFormat, totalVolume, lastSeenAt, createdAt } = propertyBuffer
+                    columns[0].push(teamId)
+                    columns[1].push(event)
+                    columns[2].push(property)
+                    columns[3].push(propertyType)
+                    columns[4].push(propertyTypeFormat)
+                    columns[5].push(totalVolume)
+                    columns[6].push(DateTime.fromSeconds(createdAt).toUTC())
+                    columns[7].push(DateTime.fromSeconds(lastSeenAt).toUTC())
+                }
+            }
+
+            // VALUES ${queryValues.join(',')}
+            await this.db.postgresQuery(
+                `INSERT INTO posthog_eventproperty(team_id, event, property, property_type, property_type_format, total_volume, created_at, last_seen_at)
+                SELECT * FROM UNNEST ($1::int[], $2::text[], $3::text[], $4::text[], $5::text[], $6::bigint[], $7::timestamp with time zone[], $8::timestamp with time zone[])
+                ON CONFLICT ON CONSTRAINT posthog_eventproperty_team_id_event_property_10910b3b_uniq DO UPDATE SET
+                    total_volume = posthog_eventproperty.total_volume + excluded.total_volume,
+                    created_at = LEAST(posthog_eventproperty.created_at, excluded.created_at),
+                    last_seen_at = GREATEST(posthog_eventproperty.last_seen_at, excluded.last_seen_at),
+                    property_type = CASE WHEN posthog_eventproperty.property_type IS NULL THEN excluded.property_type ELSE posthog_eventproperty.property_type END,
+                    property_type_format = CASE WHEN posthog_eventproperty.property_type_format IS NULL THEN excluded.property_type_format ELSE posthog_eventproperty.property_type_format END
+                `,
+                columns,
+                'eventPropertyCounterFlush'
+            )
+
+            const elapsedTime = DateTime.now().diff(startTime).as('milliseconds')
+            this.statsd?.set('flushEventPropertyCounter.Size', cacheSize)
+            this.statsd?.set('flushEventPropertyCounter.QuerySize', columns[0].length)
+            this.statsd?.timing('flushEventPropertyCounter', elapsedTime)
+            status.info(`✅ 🚽 flushEventPropertyCounter finished successfully in ${elapsedTime} ms.`)
+        } finally {
+            clearTimeout(timeout)
+        }
+    }
+}
+
+export function detectDateFormat(value: string): string | void {
+    if (value.match(/^\d{4}-\d{2}-\d{2}$/)) {
+        return 'YYYY-MM-DD'
+    }
+
+    if (value.match(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(.\d{1,3}?Z?|)$/)) {
+        return 'ISO8601 UTC'
+    }
+
+    if (value.match(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(.\d{1,3}|)\+\d{2}:\d{2}$/)) {
+        return 'ISO8601 TZ'
+    }
+}
```

Comment:
What more date formats do we need to know about? Regular unix timestamps perhaps?

Reply:
ClickHouse docs list the formats they can handle here https://clickhouse.com/docs/en/sql-reference/functions/type-conversion-functions/#parsedatetimebesteffort

Reply:
I think timestamps are pretty common too. I added a comment elsewhere, do wonder if this is required or CH is smart enough to parse them correctly when used.

Reply:
ClickHouse `parseDateTimeBestEffort is pretty clever https://github.com/PostHog/posthog/issues/6619#issuecomment-990788871

The type and format (in the short to medium term) will be most useful in the UI. E.g. if it's a timestamp and not simply a number we can display it as a date, and tell the API it is a date. See #6619